File: 2015/Day19/main.go
// Had to pull up Go for this one, wtf???
package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"strings"
)

type Replacement struct {
	from, to string
}

func main() {
	replacements, molecule := parseInput("input.txt")

	distinctMolecules := mutate(molecule, replacements)
	fmt.Printf("Part 1: %d\n", len(distinctMolecules))

	steps := search(molecule, replacements)
	fmt.Printf("Part 2: %d\n", steps)
}

func parseInput(filename string) ([]Replacement, string) {
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var replacements []Replacement
	var molecule string
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := scanner.Text()
		if strings.Contains(line, "=>") {
			parts := strings.Split(line, " => ")
			replacements = append(replacements, Replacement{parts[0], parts[1]})
		} else if line != "" {
			molecule = line
		}
	}

	return replacements, molecule
}

func mutate(molecule string, replacements []Replacement) map[string]bool {
	distinctMolecules := make(map[string]bool)
	for _, rep := range replacements {
		for i := 0; i < len(molecule); i++ {
			if strings.HasPrefix(molecule[i:], rep.from) {
				newMolecule := molecule[:i] + rep.to + molecule[i+len(rep.from):]
				distinctMolecules[newMolecule] = true
			}
		}
	}
	return distinctMolecules
}

func search(molecule string, replacements []Replacement) int {
	target := molecule
	mutations := 0

	for target != "e" {
		tmp := target
		for _, rep := range replacements {
			index := strings.Index(target, rep.to)
			if index >= 0 {
				target = target[:index] + rep.from + target[index+len(rep.to):]
				mutations++
				break
			}
		}

		if tmp == target {
			target = molecule
			mutations = 0
			shuffleReplacements(replacements)
		}
	}

	return mutations
}

func shuffleReplacements(replacements []Replacement) {
	rand.Shuffle(len(replacements), func(i, j int) {
		replacements[i], replacements[j] = replacements[j], replacements[i]
	})
}

File: 2015/Day19/main.dart
import 'dart:io';
import 'dart:math';

void main() async {
  var (replacements, molecule) = await parseInput('input.txt');

  print('Part 1: ${mutate(molecule, replacements).length}');
  print('Part 2: ${search(molecule, replacements)}');
}

Set<String> mutate(String molecule, List<Replacement> replacements) {
  var distinctMolecules = <String>{};
  for (var rep in replacements) {
    for (var i = 0; i < molecule.length; i++) {
      if (molecule.startsWith(rep.from, i)) {
        var newMolecule = molecule.substring(0, i) +
            rep.to +
            molecule.substring(i + rep.from.length);
        distinctMolecules.add(newMolecule);
      }
    }
  }
  return distinctMolecules;
}

Future<(List<Replacement>, String)> parseInput(String filename) async {
  var lines = await File(filename).readAsLines();

  var replacements = <Replacement>[];
  var molecule = '';

  for (var line in lines) {
    if (line.contains('=>')) {
      var parts = line.split(' => ');
      replacements.add(Replacement(parts[0], parts[1]));
    } else if (line.isNotEmpty) {
      molecule = line;
    }
  }

  return (replacements, molecule);
}

int search(String molecule, List<Replacement> replacements) {
  var target = molecule;
  var mutations = 0;

  while (target != 'e') {
    var tmp = target;
    for (var rep in replacements) {
      var index = target.indexOf(rep.to);
      if (index >= 0) {
        target = target.substring(0, index) +
            rep.from +
            target.substring(index + rep.to.length);
        mutations++;
        break;
      }
    }

    if (tmp == target) {
      target = molecule;
      mutations = 0;
      replacements.shuffle();
    }
  }

  return mutations;
}

void shuffleReplacementes(List<Replacement> replacements) {
  var random = new Random();
  for (var i = replacements.length - 1; i > 0; i--) {
    var j = random.nextInt(i + 1);
    var temp = replacements[i];
    replacements[i] = replacements[j];
    replacements[j] = temp;
  }
}

class Replacement {
  String from, to;
  Replacement(this.from, this.to);
}
